server/authentication_test.go

```diff
@@ -0,0 +1,134 @@
+		{"GET", healthPath, http.StatusOK, http.StatusOK},
+
+		// /debug/: server.adminServer: no auth.
+		{"GET", debugEndpoint + "vars", http.StatusOK, http.StatusOK},
+
+		// /_status/: server.statusServer: no auth.
+		{"GET", statusKeyPrefix, http.StatusOK, http.StatusOK},
+
+		// /kv/rest/: kv.RESTServer. Key cannot be empty. Response is NotFound since the key does not exist.
+		{"GET", kv.EntryPrefix + "foo", http.StatusNotFound, http.StatusUnauthorized},
+
+		// /kv/db/: kv.DBServer. These are proto reqs, but we can at least tell get past auth.
+		{"GET", kv.DBPrefix + "Get", http.StatusBadRequest, http.StatusUnauthorized},
+
+		// /ts/: ts.Server.
+		{"GET", ts.URLPrefix, http.StatusNotFound, http.StatusUnauthorized},
+	}
+
+	// HTTPS with client certs.
+	certsContext := testutils.NewTestBaseContext()
+	client, err := certsContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
+	}
+	for tcNum, tc := range testCases {
+		req, err := http.NewRequest(tc.method,
+			fmt.Sprintf("%s://%s%s", certsContext.RequestScheme(), s.ServingAddr(), tc.key),
+			nil)
+		if err != nil {
+			t.Fatalf("[%d]: error building request: %s", tcNum, err)
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			t.Fatalf("[%d]: %s: error issuing request: %s", tcNum, req.URL, err)
+		}
+
+		defer resp.Body.Close()
+		if resp.StatusCode != tc.certsStatus {
+			t.Errorf("[%d]: %s: expected status code %d, got %d", tcNum, req.URL, tc.certsStatus, resp.StatusCode)
+		}
+	}
+
+	// HTTPS without client certs.
+	noCertsContext := testutils.NewTestBaseContext()
+	noCertsContext.Certs = ""
+	client, err = noCertsContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
+	}
+	for tcNum, tc := range testCases {
+		req, err := http.NewRequest(tc.method,
+			fmt.Sprintf("%s://%s%s", noCertsContext.RequestScheme(), s.ServingAddr(), tc.key),
+			nil)
+		if err != nil {
+			t.Fatalf("[%d]: error building request: %s", tcNum, err)
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			t.Fatalf("[%d]: %s: error issuing request: %s", tcNum, req.URL, err)
+		}
+
+		defer resp.Body.Close()
+		if resp.StatusCode != tc.noCertsStatus {
+			t.Errorf("[%d]: %s: expected status code %d, got %d", tcNum, req.URL, tc.noCertsStatus, resp.StatusCode)
+		}
+	}
+
+	// Plain http.
+	insecureContext := testutils.NewTestBaseContext()
+	insecureContext.Insecure = true
+	client, err = insecureContext.GetHTTPClient()
+	if err != nil {
+		t.Fatalf("error initializing http client: %s", err)
+	}
+	for tcNum, tc := range testCases {
+		req, err := http.NewRequest(tc.method,
+			fmt.Sprintf("%s://%s%s", insecureContext.RequestScheme(), s.ServingAddr(), tc.key),
+			nil)
+		if err != nil {
+			t.Fatalf("[%d]: error building request: %s", tcNum, err)
+		}
+
+		// We're talking http to a https server. We don't even make it to a response.
+		resp, err := client.Do(req)
+		if err == nil {
+			defer resp.Body.Close()
```

Comment:
I don't think so; resp may be nil if there is an error. `golang-dev` says that `resp.Body.Close()` should only be required when err is nil: https://groups.google.com/d/msg/golang-dev/fcqo-2fgXac/vxgr5rxnFTYJ